lib/more_collections.nit

```diff
@@ -193,3 +193,332 @@ class DefaultMap[K, V]
+
+			# Plug in the new node
+			var next_node = node.next
+			insert_node(new_node, node, next_node)
+
+			# Move items at and after `index` to the new node
+			var to_displace = node.length-index
+			var offset = (new_node.capacity-to_displace)/2
+			for i in [0..to_displace[ do
+				new_node.items[offset+i] = node.items[index+i]
+				node.items[index+i] = null
+			end
+			new_node.head_index = offset
+			new_node.tail_index = offset + to_displace
+			node.tail_index -= to_displace
+
+			# Store `item`
+			if index > node.capacity / 2 then
+				new_node.items[offset-1] = item
+				new_node.head_index -= 1
+			else
+				node.items[node.head_index+index] = item
+				node.tail_index += 1
+			end
+		else
+			if node.tail_index < node.capacity then
+				# Move items towards the tail
+				node.move_tail(index, 1)
+				node.tail_index += 1
+				node.items[node.head_index + index] = item
+			else
+				# Move items towards the head
+				node.move_head(index, 1)
+				node.items[node.head_index + index-1] = item
+				node.head_index -= 1
+			end
+		end
+		length += 1
+	end
+
+	redef fun remove_at(index)
+	do
+		var node = node_at(index)
+		index = index_within_node + node.head_index
+
+		# Shift left all the elements after `index`
+		for i in [index+1 .. node.tail_index[ do
+			node.items[i-1] = node.items[i]
+		end
+		node.tail_index -= 1
+		node.items[node.tail_index] = null
+
+		length -= 1
+
+		var next_node = node.next
+		var prev_node = node.prev
+		if node.is_empty and (next_node != null or prev_node != null) then
+			# Empty and non-head or tail node, delete
+			if next_node != null then
+				next_node.prev = node.prev
+			else tail_node = prev_node.as(not null)
+
+			if prev_node != null then
+				prev_node.next = node.next
+			else head_node = next_node.as(not null)
+		end
+	end
+
+	redef fun iterator do return new UnrolledIterator[E](self)
+end
+
+# Node composing an `UnrolledList`
+#
+# Stores the elements in the `items` array. The elements in the `items` array
+# begin at `head_index` and end right before `tail_index`. The data is contiguous,
+# but there can be empty cells at the beginning and the end of the array.
+private class UnrolledNode[E]
+
+	var prev: nullable UnrolledNode[E] = null
+
+	var next: nullable UnrolledNode[E] = null
+
+	# Desired length of `items`
+	var capacity: Int
+
+	# `Array` of items in this node, filled with `null`
+	var items = new Array[nullable E].filled_with(null, capacity) is lazy
+
+	# Index of the first element in `items`
+	var head_index = 0
+
+	# Index after the last element in `items`
+	var tail_index = 0
+
+	fun length: Int do return tail_index - head_index
+
+	fun full: Bool do return length == capacity
+
+	fun is_empty: Bool do return tail_index == head_index
+
+	# Move towards the head all elements before `index` of `displace` cells
+	fun move_tail(index, displace: Int)
+	do
+		for i in [tail_index-1..head_index+index].step(-1) do
+			items[i+displace] = items[i]
```

Comment:
If I recall correctly, copy_to is implemented using `memmove`, so it should be fine right ?